### ISSUE TITLE
An exception occured when inspecting the response from SendGrid

### ISSUE THREAD

**Issue by tmacht** [Jericho/StrongGrid#269]
We just updated to the latest version of StrongGrid (0.51.0, before we used 0.50.2) and now see a lot of errors in our logs with the following message: "An exception occured when inspecting the response from SendGrid"

The error occurs when sending e-mails with the SendToSingleRecipientAsync method. The SendToSingleRecipientAsync does not throw an exception though and it does return a messageId.

**Comment by Jericho** [comment]
The log doesn't include any additional details?

If not, is there any additional information you can provide to help reproduce this problem? For instance, are you able to use Fiddler to see the request/response and provide a copy f the response here?

**Comment by Jericho** [comment]
Try 0.52.0. I made improvements to the diagnostic handler which will ensure diagnostic info will be logged in of any exception thrown when attempting to inspect the response from SendGrid. Don’t forget to set your log level to debug.

**Comment by Jericho** [comment]
Did you find additional details in your log that could help investigate?

Did you have a chance to try 0.52.0?

**Comment by tmacht** [comment]
Sorry, I had quite a busy day. See the log output below:

17:38:29.0633 Error An exception occured when inspecting the response from SendGrid
17:38:29.0633 Debug REQUEST:
  POST https://api.sendgrid.com/v3/mail/send
  accept: application/json, text/json, application/xml, text/xml, application/x-www-form-urlencoded
  StrongGrid-Diagnostic-Id: 0010cca87fb247b9adfc6514a5dbeecb
  StrongGrid-Diagnostic-Timestamp: 2585414417165
  User-Agent: StrongGrid/0.52.0, (+https://github.com/Jericho/StrongGrid)
  Authorization: ...redacted for security reasons...
  Content-Type: application/json; charset=utf-8
  Content-Length: 126001

AN EXCEPTION OCCURED: Cannot access a disposed object.
Object name: 'System.Net.Http.StreamContent'.

**Comment by Jericho** [comment]
'disposed object', that's intriguing :confused:

Couple of things that come to mind:
- I have seen similar issue when attempting to read multiple times the content of the response of a .net httpClient
- .net client disposes the content stream after reading it the first time which explain why subsequent attempts to read fail
- StrongGrid has logic to protect against this issue and allow reading the response multiple times so I'm really surprised this is happening
- Also, the 'DiagnosticHandler' is the very first StrongGrid process to read the response after it's received from SendGrid so I'm baffled that we get an error that normally occurs after the second attempt to read content.

Wild guess on my part: is it possible that you pass your own `HttpClient` or your own `HttpHanlder` to SendGrid's client constructor?

Something along the lines of:
```
var client = new SendGrid.Client("MyApiKey", myCustomHttpHandler);
```
or
```
var client = new SendGrid.Client("MyApiKey", myOwnHttpClient);
```
and, again, wild guess on my part: this custom handler has logic which reads the content of the response? This could could potentially explain why the response stream has been disposed by the time StrongGrid attempts to read it?

Anyway can you provide some of your C# code to help me reproduce? None of my tests have allowed me to observe the problem you are reporting.

**Comment by Jericho** [comment]
Wait a second, it just occured to me that this error occurs when we attempt to display the content of the REQUEST as opposed to the content of the RESPONSE. So nevermind my previous "wild guess". Continuing my investigation...

**Comment by tmacht** [comment]
No, we are neither passing a custom HttpClient nor a custom HttpHandler to StrongGrid. I can extract some of our code tomorrow morning if needed.

**Comment by Jericho** [comment]
Published 0.53.0 which contains a more robust way to log the request. Please confirm if it resolves the issue.

**Comment by tmacht** [comment]
0.53.0 seems to have resolved the issue, thanks for the quick fix!
